File: java-samples/Advanced/obligation-cordapp/workflows/src/main/java/net/corda/samples/obligation/flows/IOUTransferFlow.java
package net.corda.samples.obligation.flows;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.contracts.Command;
import net.corda.core.contracts.ContractState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.crypto.SecureHash;
import net.corda.core.flows.*;
import net.corda.core.utilities.ProgressTracker;
import net.corda.samples.obligation.contracts.IOUContract.Commands.Transfer;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.identity.AbstractParty;
import net.corda.core.identity.Party;
import net.corda.core.node.services.Vault;
import net.corda.core.node.services.vault.QueryCriteria;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.samples.obligation.contracts.IOUContract;
import net.corda.samples.obligation.states.IOUState;
import org.jetbrains.annotations.NotNull;

import java.security.PublicKey;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import static net.corda.core.contracts.ContractsDSL.requireThat;


/**
 * This is the flows which handles transfers of existing IOUs on the ledger.
 * Gathering the counterparty's signature is handled by the [CollectSignaturesFlow].
 * Notarisation (if required) and commitment to the ledger is handled by the [FinalityFlow].
 * The flows returns the [SignedTransaction] that was committed to the ledger.
 */

public class IOUTransferFlow{

    @InitiatingFlow
    @StartableByRPC
    public static class InitiatorFlow extends FlowLogic<SignedTransaction> {
        private final UniqueIdentifier stateLinearId;
        private final Party newLender;

        public InitiatorFlow(UniqueIdentifier stateLinearId, Party newLender) {
            this.stateLinearId = stateLinearId;
            this.newLender = newLender;
        }

        @Suspendable
        @Override
        public SignedTransaction call() throws FlowException {

            // 1. Retrieve the IOU State from the vault using LinearStateQueryCriteria
            List<UUID> listOfLinearIds = new ArrayList<>();
            listOfLinearIds.add(stateLinearId.getId());
            QueryCriteria queryCriteria = new QueryCriteria.LinearStateQueryCriteria(null, listOfLinearIds);

            // 2. Get a reference to the inputState data that we are going to settle.
            Vault.Page results = getServiceHub().getVaultService().queryBy(IOUState.class, queryCriteria);
            StateAndRef inputStateAndRefToTransfer = (StateAndRef) results.getStates().get(0);
            IOUState inputStateToTransfer = (IOUState) inputStateAndRefToTransfer.getState().getData();

            // 3. We should now get some of the components required for to execute the transaction
            // Here we get a reference to the default notary and instantiate a transaction builder.

            // Obtain a reference to a notary we wish to use.
            Party notary = inputStateAndRefToTransfer.getState().getNotary();

            TransactionBuilder tb = new TransactionBuilder(notary);

            // 4. Construct a transfer command to be added to the transaction.
            List<PublicKey> listOfRequiredSigners = inputStateToTransfer.getParticipants()
                    .stream().map(AbstractParty::getOwningKey)
                    .collect(Collectors.toList());
            listOfRequiredSigners.add(newLender.getOwningKey());

            Command<Transfer> command = new Command<>(
                    new Transfer(),
                    listOfRequiredSigners
            );

            // 5. Add the command to the transaction using the TransactionBuilder.
            tb.addCommand(command);

            // 6. Add input and output states to flows using the TransactionBuilder.
            tb.addInputState(inputStateAndRefToTransfer);
            tb.addOutputState(inputStateToTransfer.withNewLender(newLender), IOUContract.IOU_CONTRACT_ID);

            // 7. Ensure that this flows is being executed by the current lender.
            if (!inputStateToTransfer.lender.getOwningKey().equals(getOurIdentity().getOwningKey())) {
                throw new IllegalArgumentException("This flows must be run by the current lender.");
            }

            // 8. Verify and sign the transaction
            tb.verify(getServiceHub());
            SignedTransaction partiallySignedTransaction = getServiceHub().signInitialTransaction(tb);

            // 9. Collect all of the required signatures from other Corda nodes using the CollectSignaturesFlow
            List<FlowSession> sessions = new ArrayList<>();

            for (AbstractParty participant: inputStateToTransfer.getParticipants()) {
                Party partyToInitiateFlow = (Party) participant;
                if (!partyToInitiateFlow.getOwningKey().equals(getOurIdentity().getOwningKey())) {
                    sessions.add(initiateFlow(partyToInitiateFlow));
                }
            }
            sessions.add(initiateFlow(newLender));
            SignedTransaction fullySignedTransaction = subFlow(new CollectSignaturesFlow(partiallySignedTransaction, sessions));
            /* 10. Return the output of the FinalityFlow which sends the transaction to the notary for verification
             *     and the causes it to be persisted to the vault of appropriate nodes.
             */
            return subFlow(new FinalityFlow(fullySignedTransaction, sessions));
        }
    }


    /**
     * This is the flows which signs IOU settlements.
     * The signing is handled by the [SignTransactionFlow].
     */
    @InitiatedBy(IOUTransferFlow.InitiatorFlow.class)
    public static class Responder extends FlowLogic<SignedTransaction> {

        private final FlowSession otherPartyFlow;
        private SecureHash txWeJustSignedId;

        public Responder(FlowSession otherPartyFlow) {
            this.otherPartyFlow = otherPartyFlow;
        }

        @Suspendable
        @Override
        public SignedTransaction call() throws FlowException {
            class SignTxFlow extends SignTransactionFlow {
                private SignTxFlow(FlowSession otherPartyFlow, ProgressTracker progressTracker) {
                    super(otherPartyFlow, progressTracker);
                }

                @Override
                @NotNull
                protected void checkTransaction(SignedTransaction stx) {
                    requireThat(require -> {
                        ContractState output = stx.getTx().getOutputs().get(0).getData();
                        require.using("This must be an IOU transaction", output instanceof IOUState);
                        return null;
                    });
                    // Once the transaction has verified, initialize txWeJustSignedID variable.
                    txWeJustSignedId = stx.getId();
                }
            }

            // Create a sign transaction flows
            SignTxFlow signTxFlow = new SignTxFlow(otherPartyFlow, SignTransactionFlow.Companion.tracker());

            // Run the sign transaction flows to sign the transaction
            subFlow(signTxFlow);

            // Run the ReceiveFinalityFlow to finalize the transaction and persist it to the vault.
            return subFlow(new ReceiveFinalityFlow(otherPartyFlow, txWeJustSignedId));
        }

    }

}

File: kotlin-samples/Accounts/obligation-accounts/workflows/src/main/kotlin/net/corda/samples/obligation/flows/IOUTransferFlow.kt
package net.corda.samples.obligation.flows


import co.paralleluniverse.fibers.Suspendable
import com.r3.corda.lib.accounts.workflows.accountService
import com.r3.corda.lib.accounts.workflows.flows.RequestKeyForAccount
import com.r3.corda.lib.accounts.workflows.flows.ShareAccountInfoFlow
import com.r3.corda.lib.ci.workflows.SyncKeyMappingFlow
import com.r3.corda.lib.ci.workflows.SyncKeyMappingFlowHandler
import net.corda.core.contracts.Command
import net.corda.core.contracts.StateAndContract
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.contracts.requireThat
import net.corda.core.flows.*
import net.corda.core.identity.AbstractParty
import net.corda.core.node.services.queryBy
import net.corda.core.node.services.vault.QueryCriteria
import net.corda.core.transactions.SignedTransaction
import net.corda.core.transactions.TransactionBuilder
import net.corda.samples.obligation.contract.IOUContract
import net.corda.samples.obligation.states.IOUState
import java.util.*


/**
 * This is the flow which handles transfers of existing IOUs on the ledger.
 * Gathering the counterparty's signature is handled by the [CollectSignaturesFlow].
 * Notarisation (if required) and commitment to the ledger is handled by the [FinalityFlow].
 * The flow returns the [SignedTransaction] that was committed to the ledger.
 */
@InitiatingFlow
@StartableByRPC
class IOUTransferFlow(val linearId: UniqueIdentifier,
                      val meID: UUID,
                      val newLenderID: UUID): FlowLogic<SignedTransaction>() {
    @Suspendable
    override fun call(): SignedTransaction {

        // get the input state
        val queryCriteria = QueryCriteria.LinearStateQueryCriteria(linearId = listOf(linearId))
        val iouStateAndRef =  serviceHub.vaultService.queryBy<IOUState>(queryCriteria).states.single()
        val inputIou = iouStateAndRef.state.data

        //check if the initiator is the lender of the input IOU
        if (meID != inputIou.lenderAcctID) {
            throw IllegalArgumentException("IOU transfer can only be initiated by the IOU lender. $meID ")
        }

        //generate 3 new key pairs
        val borrowerAccount = accountService.accountInfo(inputIou.borrower.owningKey)!!.state.data
        val borrowerAcctAnonymousParty = subFlow(RequestKeyForAccount(borrowerAccount))

        val oldLenderAccount = accountService.accountInfo(meID)!!.state.data
        val oldLenderAcctPartyAndCert = subFlow(RequestKeyForAccount(oldLenderAccount))

        val newLenderAccount = accountService.accountInfo(newLenderID)!!.state.data
        val newLenderAcctAnonymousParty = subFlow(RequestKeyForAccount(newLenderAccount))

        // Build IOU output state, update participants list
        val outputIou = inputIou.withNewLender(newLenderAcctAnonymousParty, newLenderID)

        //add signers and build command
        val signers =  listOf(
            borrowerAcctAnonymousParty.owningKey,
            newLenderAcctAnonymousParty.owningKey,
            oldLenderAcctPartyAndCert.owningKey)
        val transferCommand = Command(IOUContract.Commands.Transfer(), signers)

        //build tx builder
        val notary = iouStateAndRef.state.notary
        val builder = TransactionBuilder(notary = notary)
        builder.withItems(
            iouStateAndRef,
            StateAndContract(outputIou, IOUContract.IOU_CONTRACT_ID),
            transferCommand)

        //Self verify
        builder.verify(serviceHub)

        //self sign with original lender's newly generated key
        val locallySignedTx = serviceHub.signInitialTransaction(builder,listOf(ourIdentity.owningKey,oldLenderAcctPartyAndCert.owningKey))

        val sessions = listOf(newLenderAccount.host, inputIou.borrowerHost).map {initiateFlow(it)}.toSet()
        val signedByALLParty = subFlow(CollectSignaturesFlow(locallySignedTx,sessions))
        return subFlow(FinalityFlow(signedByALLParty, sessions))

        /* The following code is to show how to collect signature manually one party at the time. */
//        //session goes to the new lender acct's parent node
//        val newLenderSession = initiateFlow(newLenderAccount.host)
//        val lenderSignature = subFlow(CollectSignatureFlow(locallySignedTx, newLenderSession, newLenderAcctAnonymousParty.owningKey))
//        val signedByNewLender = locallySignedTx.withAdditionalSignatures(lenderSignature)
//
//        //session goes to the borrower acct's parent node
//        val borrowerSession = initiateFlow(inputIou.borrowerHost)
//        val borrowerSignature = subFlow(CollectSignatureFlow(signedByNewLender, borrowerSession, borrowerAcctAnonymousParty.owningKey))
//        val signedByALLParty = signedByNewLender.withAdditionalSignatures(borrowerSignature)
//        return subFlow(FinalityFlow(signedByALLParty, listOf(newLenderSession,borrowerSession)))
    }
}

/**
 * This is the flow which signs IOU transfers.
 * The signing is handled by the [SignTransactionFlow].
 */
@InitiatedBy(IOUTransferFlow::class)
class IOUTransferFlowResponder(val flowSession: FlowSession): FlowLogic<SignedTransaction>() {
    @Suspendable
    override fun call(): SignedTransaction {
        val signedTransactionFlow = object : SignTransactionFlow(flowSession) {
            override fun checkTransaction(stx: SignedTransaction) = requireThat {
                val output = stx.tx.outputs.single().data
                "This must be an IOU transaction" using (output is IOUState)
            }
        }
        val txWeJustSignedId = subFlow(signedTransactionFlow)
        return subFlow(ReceiveFinalityFlow(otherSideSession = flowSession, expectedTxId = txWeJustSignedId.id))
    }
}